### PR TITLE
Implementation of Graph structure objects for Gremlin-Go driver.

### PR DESCRIPTION
**Pull request by lyndonbauto**
Implementation of Graph structure objects for Gremlin-Go driver.

@simonz-bq @L0Lmaker @alexey-temnikov @lyndonb-bq

### CODE REVIEW

File: gremlin-go/driver/graph.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+func (p *Property) String() string {
+	return fmt.Sprintf("p[%s->%v]", p.key, p.value)
+}
+
+func (p *Path) String() string {
+	return fmt.Sprintf("path[%s]", strings.Trim(strings.Join(strings.Fields(fmt.Sprint(p.objects)), ", "), "[]"))
+}
+
+func (p *Path) GetPathObject(key string) (error, interface{}) {
```
Comment (xiazcy) [inline gremlin-go/driver/graph.go]: I think for `go` you usually put `error` as the last item returned.

File: gremlin-go/driver/graph.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// in that it also represents an entity that is an Element and can have properties of its own.
+type VertexProperty struct {
+	Element
+	key    string // This is the label of vertex.
+	value  interface{}
+	vertex Vertex // Vertex that owns the property.
+}
+
+// Property denotes a key/value pair associated with an Edge. A property can be empty.
+type Property struct {
```
Comment (L0Lmaker) [inline gremlin-go/driver/graph.go:67]: Is there a reason why it isnt called EdgeProperty (like VertexProperty)? Comment specifies Edge, does `Property` work with both Vertex and Edges?

File: gremlin-go/driver/graph.go (hunk 1 above)
Comment (L0Lmaker) [inline gremlin-go/driver/graph.go]: Will a comment for what the function does help here or is it redundant?

File: gremlin-go/driver/graph.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+specific language governing permissions and limitations
+under the License.
+*/
+
+package gremlingo
+
+import (
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
```
Comment (simonz-bq) [inline gremlin-go/driver/graph.go]: This will probably trigger static analysis.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/graph.go (hunk 2 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graph.go:67]: This is the convention ion all the other drivers, honestly not sure why they do it this way... I think it has to do with the fact that a vertex property holds the vertex

File: gremlin-go/driver/graph.go [hunk 4]
```diff
@@ -97,14 +98,15 @@ func (p *Path) String() string {
 	return fmt.Sprintf("path[%s]", strings.Trim(strings.Join(strings.Fields(fmt.Sprint(p.objects)), ", "), "[]"))
 }
 
-func (p *Path) GetPathObject(key string) (error, interface{}) {
+// GetPathObject returns the value that corresponds to the key for the Path.
+func (p *Path) GetPathObject(key string) (interface{}, error) {
 	if len(p.objects) != len(p.labels) {
-		return errors.New("Path is invalid because it does not contain an equal number of labels and objects."), nil
+		return nil, errors.New("Path is invalid because it does not contain an equal number of labels and objects.")
```
Comment (alexey-temnikov) [inline gremlin-go/driver/graph.go:104]: Should we use a localization resource for messages?

File: gremlin-go/driver/graph.go [hunk 5]
```diff
@@ -97,14 +98,15 @@ func (p *Path) String() string {
 	return fmt.Sprintf("path[%s]", strings.Trim(strings.Join(strings.Fields(fmt.Sprint(p.objects)), ", "), "[]"))
 }
 
-func (p *Path) GetPathObject(key string) (error, interface{}) {
+// GetPathObject returns the value that corresponds to the key for the Path.
```
Comment (alexey-temnikov) [inline gremlin-go/driver/graph.go]: minor godoc nit? : `returns the value that corresponds to the key for the Path and error if the value is not present or cannot be retrieved`

File: gremlin-go/driver/graph.go (hunk 4 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graph.go:104]: I don't think our localization works with errors as of right now @simonz-bq ?

Comment (simonz-bq) [reply gremlin-go/driver/graph.go:104]: This is correct.

**Review by simonz-bq — APPROVED** [review]

**Review by xiazcy — APPROVED** [review]

File: gremlin-go/driver/graph.go (hunk 5 above)
Comment (lyndonbauto) [inline gremlin-go/driver/graph.go]:
```suggestion
// GetPathObject returns the value that corresponds to the key for the Path and error if the value is not present or cannot be retrieved.
```

**Pull request merged**